### PR TITLE
output: add nats streaming stan output

### PR DESCRIPTION
**Pull request by cpanato**
**What type of PR is this?**
/kind feature

**Any specific area of the project related to this PR?**
/area outputs

**What this PR does / why we need it**:
Add nats streaming stan to the outputs

we can refactor both nats/stan maybe later if needed.

tested with a nats/stan server

- created a kind cluster
- deployed the nats/stan using the guide: https://docs.nats.io/nats-on-kubernetes/minimal-setup

then:

sending a falco sample message:

falcosidekick logs:

```console
2020/12/09 12:49:35 [INFO]  : Enabled Outputs : STAN
2020/12/09 12:49:35 [INFO]  : Falco Sidekick is up and listening on port 2801
2020/12/09 12:49:35 [INFO]  : Debug mode : true
2020/12/09 13:02:55 [DEBUG] : Falco's payload : {"output":"This is a test from falcosidekick","priority":"Debug","rule":"Test rule","time":"2020-12-09T12:02:55Z","output_fields":null}
2020/12/09 13:02:55 [INFO]  : STAN - Publish OK
2020/12/09 13:02:56 [DEBUG] : Falco's payload : {"output":"This is a test from falcosidekick","priority":"Debug","rule":"Test rule","time":"2020-12-09T12:02:56Z","output_fields":null}
2020/12/09 13:02:57 [INFO]  : STAN - Publish OK
```

created a consumer app to read:

```console
./stan-nats

Received a message: {"output":"This is a test from falcosidekick","priority":"Debug","rule":"Test rule","time":"2020-12-09T12:03:02Z","output_fields":null}
```

**Which issue(s) this PR fixes**:

None

**Special notes for your reviewer**:

/assign @Issif

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## 2.20.0 - 2020-12-09
```
Comment (Issif) [inline CHANGELOG.md:3]: No need to manage Changelog, I do it when I release a new version. The next release will bring other improvments.

File: README.md [hunk 2]
```diff
@@ -28,6 +28,7 @@ Currently available outputs are :
 * [**Elasticsearch**](https://www.elastic.co/)
 * [**Loki**](https://grafana.com/oss/loki)
 * [**NATS**](https://nats.io/)
+* [**NATS Streaming STAN**](https://docs.nats.io/nats-streaming-concepts/intro)
```
Comment (Issif) [inline README.md:31]:
```suggestion
* [**STAN (NATS Streaming)**](https://docs.nats.io/nats-streaming-concepts/intro)
```

I think it's less ambiguous.

File: config.go [hunk 3]
```diff
@@ -95,6 +95,8 @@ func getConfig() *types.Configuration {
 	v.SetDefault("SMTP.To", "")
 	v.SetDefault("SMTP.OutputFormat", "html")
 	v.SetDefault("SMTP.MinimumPriority", "")
+	v.SetDefault("STAN.ClusterID", "")
+	v.SetDefault("STAN.ClientID", "")
```
Comment (Issif) [inline config.go:99]:
```suggestion
	v.SetDefault("STAN.HostPort", "")
	v.SetDefault("STAN.ClusterID", "")
	v.SetDefault("STAN.ClientID", "")
```

File: handlers.go [hunk 4]
```diff
@@ -172,6 +172,10 @@ func forwardEvent(falcopayload types.FalcoPayload) {
 		go natsClient.NatsPublish(falcopayload)
 	}
 
+	if config.Stan.NatsHostPort != "" && config.Stan.ClusterID != "" && config.Stan.ClientID != "" && (priorityMap[strings.ToLower(falcopayload.Priority)] >= priorityMap[strings.ToLower(config.Stan.MinimumPriority)] || falcopayload.Rule == TestRule) {
```
Comment (Issif) [inline handlers.go:175]:
```suggestion
	if config.Stan.HostPort != "" && config.Stan.ClusterID != "" && config.Stan.ClientID != "" && (priorityMap[strings.ToLower(falcopayload.Priority)] >= priorityMap[strings.ToLower(config.Stan.MinimumPriority)] || falcopayload.Rule == TestRule) {
```

File: main.go [hunk 5]
```diff
@@ -179,6 +181,18 @@ func init() {
 		}
 	}
 
+	if config.Stan.NatsHostPort != "" && config.Stan.ClusterID != "" && config.Stan.ClientID != "" {
```
Comment (Issif) [inline main.go:184]:
```suggestion
	if config.Stan.HostPort != "" && config.Stan.ClusterID != "" && config.Stan.ClientID != "" {
```

File: main.go [hunk 6]
```diff
@@ -179,6 +181,18 @@ func init() {
...
+		var err error
+		stanClient, err = outputs.NewClient("STAN", config.Stan.NatsHostPort, config, stats, promStats, statsdClient, dogstatsdClient)
+		if err != nil {
+			config.Stan.NatsHostPort = ""
```
Comment (Issif) [inline main.go:188]:
```suggestion
			config.Stan.HostPort = ""
```

File: main.go [hunk 7]
```diff
@@ -179,6 +181,18 @@ func init() {
 		}
 	}
 
+	if config.Stan.NatsHostPort != "" && config.Stan.ClusterID != "" && config.Stan.ClientID != "" {
+		var err error
+		stanClient, err = outputs.NewClient("STAN", config.Stan.NatsHostPort, config, stats, promStats, statsdClient, dogstatsdClient)
```
Comment (Issif) [inline main.go:186]:
```suggestion
		stanClient, err = outputs.NewClient("STAN", config.Stan.HostPort, config, stats, promStats, statsdClient, dogstatsdClient)
```

File: types/types.go [hunk 8]
```diff
@@ -131,6 +132,13 @@ type natsOutputConfig struct {
 	MinimumPriority string
 }
 
+type stanOutputConfig struct {
+	NatsHostPort    string
```
Comment (Issif) [inline types/types.go:136]:
```suggestion
	HostPort    string
```

For keeping a coherence with others

File: outputs/stan.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+// StanPublish publishes event to NATS Streaming
+func (c *Client) StanPublish(falcopayload types.FalcoPayload) {
+	nc, err := stan.Connect(c.Config.Stan.ClusterID, c.Config.Stan.ClientID, stan.NatsURL(c.EndpointURL.String()))
+	if err != nil {
+		go c.CountMetric("outputs", 1, []string{"output:stan", "status:error"})
+		c.Stats.Stan.Add(Error, 1)
+		c.PromStats.Outputs.With(map[string]string{"destination": "stan", "status": Error}).Inc()
+		c.Stats.Stan.Add("total", 1)
```
Comment (Issif) [inline outputs/stan.go:19]: This should be before line #14, as we increment in anycase.

**Review by Issif — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (cpanato) [reply CHANGELOG.md:3]: removed

File: outputs/stan.go (hunk 9 above)
Comment (cpanato) [reply outputs/stan.go:19]: it is in line 41, but i will make just one

File: CHANGELOG.md (hunk 1 above)
Comment (Issif) [reply CHANGELOG.md:3]: Strange, I still can see the file in the PR

Comment (cpanato) [reply CHANGELOG.md:3]: 😂 it is done but not pushed :)

File: README.md [hunk 10]
```diff
@@ -181,6 +182,12 @@ nats:
   # hostport: "" # nats://{domain or ip}:{port}, if not empty, NATS output is enabled
   # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
 
+stan:
+  # hostport: "" # nats://{domain or ip}:{port}, if not empty, NATS output is enabled
```
Comment (Issif) [inline README.md:186]:
```suggestion
  # hostport: "" # nats://{domain or ip}:{port}, if not empty, STAN output is enabled
```

File: README.md [hunk 11]
```diff
@@ -181,6 +182,12 @@ nats:
...
+  # clusterid: "" # Name of the cluster
+  # clientid: "" # Client ID to refer
```
Comment (Issif) [inline README.md:188]:
```suggestion
  # clusterid: "" # Cluster name, if not empty, STAN output is enabled
  # clientid: "" # Client ID, if not empty, STAN output is enabled
```

**Review by Issif — CHANGES_REQUESTED** [review]

File: README.md (hunk 10 above)
Comment (cpanato) [reply README.md:186]: thanks for the catches, i was blind on those

**Review by Issif — APPROVED** [review]
👍

**Comment by cpanato** [comment]
/hold cancel

tests done using both config file and env vars

**Pull request merged**